src/main/scala/viper/silver/parser/ParseAstKeyword.scala

```diff
@@ -19,6 +22,8 @@ trait LeftSpace extends PReservedString { override def leftPad = " " }
 trait RightSpace extends PReservedString { override def rightPad = " " }
 case class PReserved[+T <: PReservedString](rs: T)(val pos: (Position, Position)) extends PNode with PLeaf {
   override def display = rs.display
+  // Need to override implementation because pretty-printing will add space padding
+  override def reformat(implicit ctx: ReformatterContext): List[RNode] = rt(rs.token)
```

Comment:
I would change `leftPad` and `rightPad` to be Boolean (I think currently they only are `""` or `" "`) and then add a space to the left/right based on that. I think you can do the majority of spaces this way. You could also add a `addNewline` field which is set for e.g. all stmt keywords and `function`/`method`/`requires` etc. which adds a newline before the keyword. Lastly, I would define a some things as "grouping" in the sense that they add two spaces after a newline for all contained `RNode`s, which would probably handle most cases of indentation (I think you already do this below, but I'm not 100% sure).